_source/_posts/2025-07-28-express-oauth-pkce.md

```diff
@@ -0,0 +1,700 @@
+
+With Okta, you can implement modern authentication features and focus on your application logic without worrying about authentication infrastructure.
+
+## A secure web app using Express, OAuth 2.0, and PKCE
+
+Let's build an expense dashboard where users log in with Okta and view spending data based on their role. Whether they work in Finance, Marketing, or HR, each team views only its own expenses. To keep things minimal in this demo project, we'll define roles and users directly in the app.
+
+You'll also use OpenID Connect (OIDC) through the openid-client library for authentication. Then, you'll map each user's email from the ID token to a team. The dashboard applies principles of least privilege and displays expenses by team, so each user sees only their department's spending.
+
+**Prerequisites**
+
+* Node.js installed (v22+ recommended)
+
+* [Okta Integrator Free Plan org](https://developer.okta.com/signup/)
+
+## Create your Express project and install dependencies
+
+Create a new project folder named `express-project-okta`, and open a terminal window in the project folder.
+
+Initialize a new Node.js project:
+
+```sh
+npm init -y
+```
+
+Install the required packages:
+
+```sh
+npm install express@5.1 passport@0.7 openid-client@6.6 express-session@1.18 ejs@3.1 express-ejs-layouts@2.5 dotenv
+```
+
+
+Now, install the development dependencies:
+
+```sh
+npm install --save-dev nodemon
+``` 
+
+**What do these dependencies do?** 
+
+These installed packages become your Express project's dependencies.
+
+* **`express`**: Handles routing and HTTP middleware for your web app
+
+* **`passport`**: Sets up and maintains server-side sessions
+
+* **`openid-client`**: Node.js OIDC library with PKCE support; handles the OAuth handshake and token exchange.
+
+* **`express-session`**: Manages user sessions on the server
+
+* **`dotenv`**: Loads environment variables from a `.env` file
+
+* **`ejs`**: Enables dynamic HTML rendering using embedded JavaScript templates
+
+* **`express-ejs-layouts`**: Adds layout support to EJS, helping manage common layout structures across views
+
+## Configure environment variables for OIDC authentication
+
+Create a `.env` file in the root directory with placeholders for your Okta configuration.
+
+```
+OKTA_ISSUER=https://{yourOktaDomain}/oauth2/default 
```

Comment:
```suggestion
OKTA_ISSUER=https://{yourOktaDomain}
```

Are you using the custom authorization server? if not, then this is the correct URL